### PR TITLE
B/interfaces

### PR DESCRIPTION
**Pull request by worldofjoni**
# Pull Request for initial Implementation

## Backend
Make sure, you have done de following before merging this pull request:
- [x] Your code is **documented** using doc comments.
- [x] You added appropriate **logging** in your code.
- [x] You added **unit tests** to all your implemented functions.
- [ ] You ran `cargo format` to format your code.
- [ ] You ran `cargo clippy` to check for improvements.
- [x] You added the **class descriptions** of the classes you implemented in the **latex document** on overleaf.
      Make sure to mark any changes you made to the design documents specification.

### CODE REVIEW

**Comment by worldofjoni** [comment]
@Whatsuup is it intetnional that you included the `.idea` folder?

**Review by Whatsuup — COMMENTED** [review]
.idea folders deleted. Cargo clippy and format worked?

File: backend/src/interface/image_hoster/model/mod.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    pub id: String,
+    /// The url which short links to the image. Different to the url provided by an user.
+    pub image_url: String,
+    /// The licence of the image.
+    pub licence: String,
+}
+
+/// Enum describing the possible ways, a image hoster request can fail.
+#[derive(Debug, Error)]
+pub enum ImageHosterError {
```
Comment (worldofjoni) [inline backend/src/interface/image_hoster/model/mod.rs:17]: Do we really need all these detailed error messages?

File: backend/src/interface/image_hoster/model/mod.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    InvalidApiKey,
+    /// Service currently unavailable error
+    #[error("the requested service is temporarily unavailable")]
+    ServiceUnavailable,
+    ///  Write operation failed error
+    #[error("the requested operation failed due to a temporary issue")]
+    WriteFailed,
+    /// Format "xxx" not found error
+    #[error("the requested response format was not found")]
+    FormatNotFound(#[from] Box<dyn Error>),
```
Comment (worldofjoni) [inline backend/src/interface/image_hoster/model/mod.rs:35]: Which error type should get inserted here?

Comment (worldofjoni) [reply backend/src/interface/image_hoster/model/mod.rs:35]: Additionally, how are errors handeld when no http connection could be established etc?

File: backend/src/interface/image_review/mod.rs [hunk 3]
```diff
@@ -1 +1,10 @@
 //! This interface allows starting the process of checking the existence of images linked to meals.
+
+use async_trait::async_trait;
+
+/// This interface allows starting the image review process to check for deleted images at the image hoster.
+#[async_trait]
+pub trait ImageReviewScheduling {
```
Comment (worldofjoni) [inline backend/src/interface/image_review/mod.rs:7]: Here, clippy wants us to remove package name redundancy in the trait name. Should we follow or ignore that?

File: backend/src/interface/mealplan_management/mod.rs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+//! This interface allows starting the operations for updating the menu from the the canteen's website.
+
+pub trait MensaParseScheduling {
```
Comment (worldofjoni) [inline backend/src/interface/mealplan_management/mod.rs:3]: I think these methods should also be async.

File: backend/src/interface/mensa_parser/mod.rs [hunk 5]
```diff
@@ -1 +1,15 @@
...
+
+mod model;
+
+#[async_trait]
+/// Parser interface. Provides functions which return canteen structs. Canteen structs contain raw data obtained by parsing mealplans.
+pub trait MealplanParser {
+    /// Initiate a parse procedure. Returns a canteen struct containing mealplan data of the given date.
+    async fn parse(day: Date) -> Vec<ParseCanteen>;
+    /// Initiate a parse procedure. Returns a tuple containing mealplan data of the next four weeks. The tuple contains a canteen struct with the related date.
+    async fn parse_all() -> (Date, Vec<ParseCanteen>);
```
Comment (worldofjoni) [inline backend/src/interface/mensa_parser/mod.rs:14]: Shouldnt there be also multiple dates returned?

File: backend/src/interface/mensa_parser/model/mod.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
+//! These structs are used for parse operations
```
Comment (worldofjoni) [inline backend/src/interface/mensa_parser/model/mod.rs:1]: Punctuation.

File: backend/src/interface/mensa_parser/model/mod.rs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    /// Name of the dish.
+    pub name: String,
+    /// Price of the meal for students, employees, guests and pupils.
+    pub price: Price,
+    /// All containing allergens.
+    pub allergens: Vec<Allergen>,
+    /// All containing additives.
+    pub additives: Vec<Additive>,
+    /// Meal-Type of the dish.
+    pub c_type: MealType,
```
Comment (worldofjoni) [inline backend/src/interface/mensa_parser/model/mod.rs:32]: In doc comment: Either reference the `MealType` enum or write it as "Meal type of ..." in correct english orthography.

File: backend/src/interface/persistent_data/mod.rs [hunk 8]
```diff
@@ -1 +1,141 @@
 //! The interfaces specified here allow access to data stored in a persistent datastore like a database.
+use crate::interface::persistent_data::model::{
+    ApiKey, Canteen, DataError, Image, ImageInfo, Line, Meal, Side,
+};
```
Comment (worldofjoni) [inline backend/src/interface/persistent_data/mod.rs:4]: maybe its prettyer to import `model::*`?
Depends on what the formater would do with that...

File: backend/src/interface/persistent_data/mod.rs [hunk 9]
```diff
@@ -1 +1,141 @@
...
+    /// Adds a new side entity to the database. Returns the new entity.
+    async fn insert_side(
+        name: String,
+        c_type: MealType,
+        price: Price,
+        next_served: Date,
+        allergens: Vec<Allergen>,
+        additives: Vec<Additive>,
+    ) -> Result<Side>;
+}
```
Comment (worldofjoni) [inline backend/src/interface/persistent_data/mod.rs:70]: I would suggest putting all these Traits into seperate files as they re quite long.

File: backend/src/interface/mensa_parser/model/mod.rs (hunk 7 above)
Comment (worldofjoni) [reply backend/src/interface/mensa_parser/model/mod.rs:32]: Also: what does the `c´ in `c_type` stand for?
I think such abrevations shout be avoided.

File: backend/src/interface/persistent_data/model/mod.rs [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+use std::error::Error;
+
+use std::string::String;
+use thiserror::Error;
+use util::MealType;
+use uuid::Uuid;
+
+/// Enumerations for possible data request faults
+#[derive(Debug, Error)]
+pub enum DataError {
```
Comment (worldofjoni) [inline backend/src/interface/persistent_data/model/mod.rs:13]: Maybe put the error inside the `persistent_datamod.rs` directly, as it is not part of the model?

File: backend/src/interface/persistent_data/model/mod.rs [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    /// Price of the side for students, employees, guests and pupils.
+    pub price: Price,
+    /// All containing allergens.
+    pub allergens: Vec<Allergen>,
+    /// All containing additives.
+    pub additives: Vec<Additive>,
+}
+
+/// This structure is used for database operations. This image structure is based on the database entity 'image'.
+pub struct Image {
```
Comment (worldofjoni) [inline backend/src/interface/persistent_data/model/mod.rs:95]: Should we merge `Image` and `ImageInfo`?

File: backend/src/util/mod.rs [hunk 12]
```diff
@@ -1 +1,146 @@
...
+    ViolatesRights,
+    /// This picture should be removed for some other reason.
+    Other,
+}
+
+/// Date type used in multiple paces.
+pub type Date = NaiveDate;
+
+/// This struct contains all price classes
+pub struct Price {
```
Comment (worldofjoni) [inline backend/src/util/mod.rs:137]: Maybe specify in docs that prices are in (euro) cents?

File: backend/src/util/mod.rs [hunk 13]
```diff
@@ -1 +1,146 @@
 //! This package contains structures that are used in many components.
+//!
+//! For a complete list and explanations you can see [here](https://www.sw-ka.de/media/?file=4458listeallergesetzlichausweisungspflichtigenzusatzstoffeundallergenefuerwebsite160218.pdf&download).
+
+use chrono::NaiveDate;
+
+/// This enum lists every possible allergen a meal can have.
+pub enum Allergen {
```
Comment (worldofjoni) [inline backend/src/util/mod.rs:8]: Should we spread them also over multiple files?

File: backend/src/interface/image_hoster/model/mod.rs (hunk 2 above)
Comment (Whatsuup) [reply backend/src/interface/image_hoster/model/mod.rs:35]: - When a formatting error occurs, it is important to know what formatting caused the error.
- An additional error type if "no http connection could be established". Something like "NotConnected".

File: backend/src/interface/image_review/mod.rs (hunk 3 above)
Comment (Whatsuup) [reply backend/src/interface/image_review/mod.rs:7]: Could not suppress the error

File: backend/src/interface/mealplan_management/mod.rs (hunk 4 above)
Comment (Whatsuup) [reply backend/src/interface/mealplan_management/mod.rs:3]: No, the functions have no returns

File: backend/src/interface/mensa_parser/mod.rs (hunk 5 above)
Comment (Whatsuup) [reply backend/src/interface/mensa_parser/mod.rs:14]: Yes, kind of.  Vec<(Date, Vec<ParseCanteen>)>

File: backend/src/interface/mensa_parser/model/mod.rs (hunk 6 above)
Comment (Whatsuup) [reply backend/src/interface/mensa_parser/model/mod.rs:1]: .

File: backend/src/interface/mensa_parser/model/mod.rs (hunk 7 above)
Comment (Whatsuup) [reply backend/src/interface/mensa_parser/model/mod.rs:32]: c_type => custom_type. Alternate of meal_type. Term `type` is used for rust syntax.

File: backend/src/interface/persistent_data/mod.rs (hunk 8 above)
Comment (Whatsuup) [reply backend/src/interface/persistent_data/mod.rs:4]: Idk if this is really important. But if clippy wants to, we can change it.

File: backend/src/interface/persistent_data/mod.rs (hunk 9 above)
Comment (Whatsuup) [reply backend/src/interface/persistent_data/mod.rs:70]: I thought about that, but if we do it, then we do it everywhere. We don't want inconsistent folder/file styles.

File: backend/src/interface/persistent_data/model/mod.rs (hunk 10 above)
Comment (Whatsuup) [reply backend/src/interface/persistent_data/model/mod.rs:13]: I can't see the advantage of that. The idea behind this structure: Interfaces into root, Structs/Enums into root/model.

File: backend/src/util/mod.rs (hunk 13 above)
Comment (Whatsuup) [reply backend/src/util/mod.rs:8]: see above

File: backend/src/util/mod.rs (hunk 12 above)
Comment (Whatsuup) [reply backend/src/util/mod.rs:137]: maybe yes

File: backend/src/interface/image_hoster/model/mod.rs (hunk 1 above)
Comment (Whatsuup) [reply backend/src/interface/image_hoster/model/mod.rs:17]: No, but I already created them. Some of them could be merged, to minimize this enum.

File: backend/src/interface/admin_notification/mod.rs [hunk 14]
```diff
@@ -1 +1,33 @@
...
+use async_trait::async_trait;
+use uuid::Uuid;
+
+use crate::util::ReportReason;
+
+/// Interface for notification of administrators.
+#[async_trait]
+pub trait AdminNotification {
+    /// Notifies an administrator about a newly reported image and the response automatically taken.
+    async fn notify_admin_image_report(info: ImageReportInfo);
```
Comment (Whatsuup) [inline backend/src/interface/admin_notification/mod.rs:12]: Async not necessary for functions without return?(!)

File: backend/src/interface/persistent_data/model/mod.rs (hunk 11 above)
Comment (worldofjoni) [reply backend/src/interface/persistent_data/model/mod.rs:95]: ok

File: backend/src/interface/persistent_data/model/mod.rs (hunk 10 above)
Comment (worldofjoni) [reply backend/src/interface/persistent_data/model/mod.rs:13]: For me, model are just data models. The error type has a more universal role imo.

File: backend/src/interface/admin_notification/mod.rs (hunk 14 above)
Comment (worldofjoni) [reply backend/src/interface/admin_notification/mod.rs:12]: Functions without an explicit return type return `()`. You cannot `await` an async function in a synchronous one.

File: backend/src/interface/mealplan_management/mod.rs (hunk 4 above)
Comment (worldofjoni) [reply backend/src/interface/mealplan_management/mod.rs:3]: In rust, a function without explicit return type returns `()`. You cannot await a async function in a synchronous one.

**Comment by worldofjoni** [comment]
We also need to add the two new enum variants, see discord.

File: backend/src/interface/mensa_parser/model/mod.rs (hunk 7 above)
Comment (worldofjoni) [reply backend/src/interface/mensa_parser/model/mod.rs:32]: Im not happy with that... can we use the full version `meal_type`?

**Pull request merged**